libs/model/src/services/tokenBalanceCache/providers/get_spl_balances.ts

```diff
@@ -0,0 +1,102 @@
+import { logger } from '@hicommonwealth/logging';
+import { getAccount, getAssociatedTokenAddressSync } from '@solana/spl-token';
+import { Connection, PublicKey } from '@solana/web3.js';
+import { ChainNodeInstance } from 'model/src/models/chain_node';
+import { fileURLToPath } from 'url';
+import { Balances, GetSPLBalancesOptions } from '../types';
+import { failingChainNodeError } from '../util';
+
+const __filename = fileURLToPath(import.meta.url);
+const log = logger(__filename);
+
+export async function __get_spl_balances(
+  chainNode: ChainNodeInstance,
+  options: GetSPLBalancesOptions,
+): Promise<Balances> {
+  const rpcEndpoint = chainNode.private_url || chainNode.url;
+  if (options.addresses.length === 1) {
+    const balances: Balances = {};
+    balances[options.addresses[0]] = await getSingleSPLBalance(
+      new Connection(rpcEndpoint),
+      options.mintAddress,
+      options.addresses[0],
+    );
+    return balances;
+  }
+
+  return (
+    await getBulkSPLBalances(
+      rpcEndpoint,
+      options.mintAddress,
+      options.addresses,
+      options.batchSize,
+    )
+  ).balances;
+}
+
+async function getBulkSPLBalances(
+  rpcEndpoint: string,
+  mintAddress: string,
+  addresses: string[],
+  batchSize = 100,
+): Promise<{ balances: Balances; failedAddresses: string[] }> {
+  const connection = new Connection(rpcEndpoint);
+  let batchRequestPromises = [];
+  const idAddressMap: Balances = {};
+  const balances: Balances = {};
+  let failedAddresses: string[] = [];
+  let id = 0;
+  for (
+    let startIndex = 0;
+    startIndex < addresses.length;
+    startIndex += batchSize
+  ) {
+    const endIndex = Math.min(startIndex + batchSize, addresses.length);
+    const batchAddresses = addresses.slice(startIndex, endIndex);
+    for (const address of batchAddresses) {
+      batchRequestPromises.push(
+        getSingleSPLBalance(connection, mintAddress, address),
```

Comment:
I think we can optimize this if we use `getMultipleAccounts` when `addresses.length > 1`.

Reply:
I experimented with getMultipleAccounts and can explore further but the issue is that if any of the accounts passed do not exist which on solana is the case for users who havent interacted with the token yet or have closed their token account - the call fails. 

If you check out the first step in `getSingleSPLBalance` we are running a ATA(associate token account - on sol a wallet owns an ATA for every SPL token it holds) calculation. ATA addresses are deterministic based on wallet address and token mint address so we can always calculate them, but that doesnt guarantee they exist on chain so upon calling getAccount or including in getMultipleAccount with a non-existent token account we get an error which tbh from the RPC perspective is pretty annoying its not just null or something. 

Here's an example fwiw 

The first address has tokens, second address doenst have tokens 

<img width="722" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/36428666/45feed31-358d-45ad-9075-cc19bb4d458b">

upon hitting the cache the whole getMultipleAccounts call fails:
<img width="722" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/36428666/777aea62-dd02-4999-9099-4b8342ab00ff">

Obv the next idea would be to include only existing ATA's but that obviously still requires an on-chain check for existing accounts